File: Python/word-ladder.py
# Time:  O(26 * d * l) = O(d * l), d is the size of wordlist, l is the max length of words
# Space: O(d * l)

from string import ascii_lowercase


# two-end bfs
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        lookup = set(wordList)
        if endWord not in lookup:
            return 0
        lookup.remove(endWord)
        ladder = 2
        left, right = set([beginWord]), set([endWord])
        while left and right:
            if len(left) > len(right):
                left, right = right, left
            new_left = []
            for word in left:
                for i in xrange(len(word)):
                    for j in ascii_lowercase:
                        new_word = word[:i] + j + word[i+1:]
                        if new_word in right:
                            return ladder
                        if new_word in lookup:
                            lookup.remove(new_word)
                            new_left.append(new_word)
            left = new_left
            ladder += 1
        return 0


# Time:  O(26 * d * l) = O(d * l), d is the size of wordlist, l is the max length of words
# Space: O(d * l)
class Solution2(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        lookup = set(wordList)
        if endWord not in lookup:
            return 0
        ladder = 2
        q = [beginWord]
        while q:
            new_q = []
            for word in q:
                for i in xrange(len(word)):
                    for j in ascii_lowercase:
                        new_word = word[:i] + j + word[i+1:]
                        if new_word == endWord:
                            return ladder
                        if new_word in lookup:
                            lookup.remove(new_word)
                            new_q.append(new_word)
            q = new_q
            ladder += 1
        return 0

File: C++/word-ladder.cpp
// Time:  O(b^(d/2)), b is the branch factor of bfs, d is the result depth
// Space: O(w * l), w is the number of words, l is the max length of words

// two-end bfs
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> words(cbegin(wordList), cend(wordList));
        if (!words.count(endWord)) {
            return 0;
        }
        int ladder = 2;
        unordered_set<string> left = {beginWord}, right = {endWord};
        while (!empty(left)) {
            for (const auto& word : left) {
                words.erase(word);
            }
            unordered_set<string> new_left;
            for (const auto& word : left) {
                auto new_word = word;
                for (int i = 0; i < size(new_word); ++i) {
                    char prev = new_word[i];
                    for (int j = 0; j < 26; ++j) {
                        new_word[i] = 'a' + j;
                        if (!words.count(new_word)) {
                            continue;
                        }
                        if (right.count(new_word)) {
                            return ladder;
                        }
                        new_left.emplace(new_word);
                    }
                    new_word[i] = prev;
                }
            }
            left = move(new_left);
            ++ladder;
            if (size(left) > size(right)) {
                swap(left, right);
            }
        }
        return 0;
    }
};

// Time:  O(b^d), b is the branch factor of bfs, d is the result depth
// Space: O(w * l), w is the number of words, l is the max length of words
class Solution2 {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> lookup(cbegin(wordList), cend(wordList));
        if (!lookup.count(endWord)) {
            return 0;
        }
        int ladder = 2;
        for (vector<string> q = {beginWord}; !q.empty(); ++ladder) {
            vector<string> new_q;
            for (const auto& word : q) {
                auto new_word = word;
                for (int i = 0; i < new_word.length(); ++i) {
                    char prev = new_word[i];
                    for (int j = 0; j < 26; ++j) {
                        new_word[i] = 'a' + j;
                        if (new_word == endWord) {
                            return ladder;
                        }
                        if (lookup.count(new_word)) {
                            lookup.erase(new_word);
                            new_q.emplace_back(new_word);
                        }
                    }
                    new_word[i] = prev;
                }
            }
            q = move(new_q);
        }
        return 0;
    }
};
